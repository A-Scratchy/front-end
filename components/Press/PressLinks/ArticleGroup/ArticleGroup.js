import React, { Component } from 'react';
import PropTypes from 'prop-types';
import OutboundLink from 'components/_common_/OutboundLink/OutboundLink';
import Button from 'components/_common_/Button/Button';
import styles from './ArticleGroup.css';

class ArticleGroup extends Component {
  static propTypes = {
    articles: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string.isRequired,
        url: PropTypes.string.isRequired,
      }),
    ).isRequired,
    numberOfInitiallyVisibleLinks: PropTypes.number.isRequired,
    region: PropTypes.string.isRequired,
  };

  state = {
    areAllLinksVisible: false,
  };

  clickHandler = () => {
    this.setState(prevState => ({ areAllLinksVisible: !prevState.areAllLinksVisible }));
  };

  render() {
    const { areAllLinksVisible } = this.state;
    const { articles, region, numberOfInitiallyVisibleLinks } = this.props;
    return (
      <div className={styles.articlesGroup}>
        <h2>{region}</h2>
        <ul>
          {articles.map((link, index) => {
            if (index >= numberOfInitiallyVisibleLinks && !areAllLinksVisible) {
              return null;
            }
            return (
              <li key={`GroupLink_${link.url}`}>
                <OutboundLink href={link.url}>{link.title}</OutboundLink>
              </li>
            );
          })}
        </ul>
        {articles.length > numberOfInitiallyVisibleLinks && (
          <Button // aria needs to be passed and accepted by the button component
            aria-pressed={areAllLinksVisible} // eslint-disable-line
            className={styles.areAllLinksVisibleButton}
            theme={areAllLinksVisible ? 'slate' : 'primary'}
            onClick={this.clickHandler}
          >
            {areAllLinksVisible ? 'Show Less' : 'Show All'}
          </Button>
        )}
      </div>
    );
  }
}

export default ArticleGroup;
