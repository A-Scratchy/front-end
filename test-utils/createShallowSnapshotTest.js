/* eslint-env jest */
import ShallowRenderer from 'react-test-renderer/shallow';

/**
 * Used to create a snapshot test off of surface leaves of the tree.
 * Useful for regression tests on containers or larger components.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const shallowRenderer = new ShallowRenderer();
  const tree = shallowRenderer.render(Component);

  expect(tree).toMatchSnapshot();
};
