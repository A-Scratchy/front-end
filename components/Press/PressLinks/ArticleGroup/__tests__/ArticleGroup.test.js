import React from 'react';
import createShallowSnapshotTest from 'test-utils/createShallowSnapshotTest';
import { mount, shallow } from 'enzyme';

import ArticleGroup from '../ArticleGroup';

describe('ArticleGroup', () => {
  test('should render properly with required props', () =>
    createShallowSnapshotTest(
      <ArticleGroup
        title="test"
        links={[{ title: 'Example', url: 'https://example.com' }]}
        numberOfInitiallyVisibleLinks={1}
      />,
    ));

  test('should render properly with required props and 3 links and a button', () =>
    createShallowSnapshotTest(
      <ArticleGroup
        title="test"
        links={[
          { title: 'Example', url: 'https://example.com' },
          { title: 'Example', url: 'https://example.com' },
          { title: 'Example', url: 'https://example.com' },
        ]}
        numberOfInitiallyVisibleLinks={1}
      />,
    ));

  test('should setState when clicking Show All button', () => {
    const ArticleGroupShallowInstance = shallow(
      <ArticleGroup
        title="test"
        links={[
          { title: 'Example', url: 'https://example.com' },
          { title: 'Example', url: 'https://example.com' },
          { title: 'Example', url: 'https://example.com' },
        ]}
        numberOfInitiallyVisibleLinks={1}
      />,
    );

    ArticleGroupShallowInstance.instance().clickHandler();

    expect(ArticleGroupShallowInstance.state().areAllLinksVisible).toEqual(true);
  });

  test('should not create a button if not enough links', () => {
    const wrap = mount(
      <ArticleGroup
        title="test"
        links={[{ title: 'Example', url: 'https://example.com' }]}
        numberOfInitiallyVisibleLinks={5}
      />,
    );

    expect(wrap.find('button').exists()).toEqual(false);
  });

  test('should create a button if enough links are available', () => {
    const wrap = mount(
      <ArticleGroup
        title="test"
        links={[
          { title: 'Example', url: 'https://example.com' },
          { title: 'Example', url: 'https://example.com' },
          { title: 'Example', url: 'https://example.com' },
        ]}
        numberOfInitiallyVisibleLinks={1}
      />,
    );

    expect(wrap.find('button').exists()).toEqual(true);
  });
});
