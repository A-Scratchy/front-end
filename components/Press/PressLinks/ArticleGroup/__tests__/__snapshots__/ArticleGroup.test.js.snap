// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleGroup should render with required props 1`] = `
<div
  className="articlesGroup"
>
  <h2>
    test
  </h2>
  <ul>
    <li>
      <OutboundLink
        analyticsEventLabel="Press Article"
        hasIcon={true}
        href="https://example.com"
        router={
          Object {
            "articles": Array [
              Object {
                "title": "Example",
                "url": "https://example.com",
              },
            ],
            "numberOfInitiallyVisibleLinks": 1,
            "region": "test",
          }
        }
      >
        Example
      </OutboundLink>
    </li>
  </ul>
</div>
`;

exports[`ArticleGroup should render with required props and 3 links and a button 1`] = `
<div
  className="articlesGroup"
>
  <h2>
    test
  </h2>
  <ul>
    <li>
      <OutboundLink
        analyticsEventLabel="Press Article"
        hasIcon={true}
        href="https://example.com"
        router={
          Object {
            "articles": Array [
              Object {
                "title": "Example",
                "url": "https://example.com",
              },
              Object {
                "title": "Example",
                "url": "https://example.com",
              },
              Object {
                "title": "Example",
                "url": "https://example.com",
              },
            ],
            "numberOfInitiallyVisibleLinks": 1,
            "region": "test",
          }
        }
      >
        Example
      </OutboundLink>
    </li>
  </ul>
  <Button
    analyticsObject={
      Object {
        "action": "Button Selected",
        "category": "Interactions",
      }
    }
    aria-pressed={false}
    className="areAllLinksVisibleButton"
    datum=""
    disabled={false}
    fullWidth={false}
    onClick={[Function]}
    tabIndex={0}
    theme="primary"
    type="button"
  >
    Show All
  </Button>
</div>
`;
